### PR TITLE
#492 Typescript migration to send

### PR DESCRIPTION
**Pull request by drewbrokke**
Hey @jwu910! Thought I'd give this one a shot. I gave all the features a quick test and it seems to be working well, but please feel free to pick it apart and double-check everything. Let me know if you have any questions or concerns. Thanks!

## Description

This migrates the remaining `js` files to use Typescript. It also adds a dedicated type file.

#492

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update
- [x] Refactor

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] My commits follow the [commitizen](https://github.com/commitizen/cz-cli#philosophy) commit convention
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: README.md
Comment (rahulk94) [inline README.md:90]: Propose using `npm start` so to avoid declaring the app start script twice.
```suggestion
npm start
```

File: src/app.ts
Comment (rahulk94) [inline src/app.ts:230]: It feels like using `(error as Error).message` everywhere you've got `error as string` is a more correct Typescript way of doing this. I'd suggest tweaking to that (you've already done that in `state.ts` so would make things more consistent).

File: src/app.ts
Comment (rahulk94) [inline src/app.ts:285]: This looks like a fun one to solve in TS 😂 I couldn't find how/where `this` is handled

**Review by rahulk94 — COMMENTED** [review]
Hey @jwu910 @drewbrokke whilst not a maintainer of this project, I've given it a review and tried it locally too. Few minor comments but looks like a good step forward in the TS migration project 🙌🏾

File: README.md
Comment (drewbrokke) [reply README.md:90]: @rahulk94 good suggestion. Looks like `npm start` already present in the readme [here](https://github.com/jwu910/check-it-out/blob/typescript-migration/README.md?plain=1#L94), so this would be a straight removal. Would you be willing to file an issue for this? It would apply just as well to the main branch as to this one.

File: src/app.ts
Comment (drewbrokke) [reply src/app.ts:230]: @rahulk94 I fully agree. 

However, I did play around with changing the git api to reject with actual `Error` types, but decided against it as that is not the intent of this effort.

IMO we should add that change under a separate issue, but I'll defer to @jwu910 for the final verdict.

File: src/app.ts
Comment (drewbrokke) [reply src/app.ts:285]: I _think_ `this` is supposed to refer to the `branchTable` variable which is of type `Widgets.ListTableElement` from `blessed`. However, this type doesn't actually publish the `items` or `selected` properties, so we'll probably just have to make our own type since we _know_ it's there.

I'll give it a shot and push an update.

**Comment by drewbrokke** [comment]
Thanks for the suggestions @rahulk94! I've left some responses, please let me know what you think.

File: README.md
Comment (rahulk94) [reply README.md:90]: Yep sure I'll open an issue + PR

Comment (jwu910) [reply README.md:90]: Oh good point. We can link npm start to the build assets. 👌

File: src/app.ts
Comment (jwu910) [reply src/app.ts:230]: It's been a long time, but if I'm remembering this part is really just for notifying events in the log, not entirely sure it will be absolutely necessary at least in this space to catch the `Error` type since it's just a message to the user.

But maybe we can change the variable name to errorMessage in the future to be more clear

File: README.md
Comment (rahulk94) [reply README.md:90]: https://github.com/jwu910/check-it-out/pull/533

File: src/app.ts
Comment (rahulk94) [reply src/app.ts:285]: Thanks - I'm okay with TS ignore for it or deferring till later if @jwu910 is.

Comment (drewbrokke) [reply src/app.ts]: Just pushed an update, take a look and lmk what you think.

**Comment by jwu910** [comment]
@rahulk94 of course, I appreciate the engagement, it has inspired @drewbrokke to come out of hiding!

Thanks Drew for the PR I'll try to give it a more thorough review this weekend or later this evening, but so far eveything looked great on first pass

File: README.md
Comment (jwu910) [reply README.md:90]: Thanks for raising this, took a look at concluded it's not really an issue. Readme offers flexibility and not forces one method over the other. We'll leave this in for now 👌🏻

**Review by jwu910 — APPROVED** [review]
This looks good @drewbrokke just merged a PR that removed some old dependencies that were no longer used. You'll have to update your package lock to deal with the conflicts after the rebase.

Appreciate the time!! CI should be fixed to run on fork PRs as well too

**Comment by drewbrokke** [comment]
Thanks @jwu910! Feel free to drop the package-lock.json commit and regen whenever you decide to merge. Otherwise, I'll update this branch tomorrow.

**Comment by drewbrokke** [comment]
@jwu910 just force-pushed.

**Review by jwu910 — APPROVED** [review]

**Comment by jwu910** [comment]
Thanksss

**Pull request merged**
